### PR TITLE
feat(enhancement) Expand Flotsam Collection Setting

### PR DESCRIPTION
**Pull request by OcelotWalrus**
# Feature Details

The Pull Request adds a new setting to the settings panel: the flotsam collection setting. It is in the 'Gameplay' category right after the 'expend ammo' setting (This pull request also removes the 'flagship flotsam collection' this is already in the game'):

**In these images it says 'flotsam pick-up' but it's now called 'flotsam collection'.**

| Parameter | Image                                                                                                     | Description                                                 |
|-----------|-----------------------------------------------------------------------------------------------------------|-------------------------------------------------------------|
| ON        | image | All your fleet can pick up flotsam                          |
| FLAGSHIP  | image | Only your flagship can pick up flotsam                      |
| ESCORT    | image | Only your escort and not your fflagship can pick up flotsam |
| OFF       | image | None of your fleet can pick up flotsam                      |

### CODE REVIEW

**Comment by OcelotWalrus** [comment]
> I prefer having a unified flotsam collection setting and here's why:
> 
>     1. It can be confusing when you have for example flagship on and fleet off (this could be solved by renaming the preference from "fleet" to "escorts").
> 
>     2. If we combined these settings, we could avoid bloating the settings panel too much (similarly to how we have one setting for autofire).

Sure it seems better. So can I'll update my mode to make so its in one single setting.

**Comment by OcelotWalrus** [comment]
See edit of https://github.com/endless-sky/endless-sky/pull/9007#issue-1791689394

**Comment by OcelotWalrus** [comment]
> I prefer having a unified flotsam collection setting and here's why:
> 
>     1. It can be confusing when you have for example flagship on and fleet off (this could be solved by renaming the preference from "fleet" to "escorts").
> 
>     2. If we combined these settings, we could avoid bloating the settings panel too much (similarly to how we have one setting for autofire).

I finished.

**Review by TomGoodIdea — COMMENTED** [review]
Thanks for making it a single setting.

It's good to maintain the same order of settings in UI and code. Or at least don't put flotsam-related settings between autoaim and autofire.

File: source/Engine.cpp [hunk 1]
```diff
@@ -2261,7 +2261,12 @@ void Engine::DoCollection(Flotsam &flotsam)
...
 	if(!collector)
 		return;
-	if(collector == player.Flagship() && !Preferences::Has("Flagship flotsam collection"))
+	const auto flotsamSetting = Preferences::GetFlotsam();
+	// If the collector is the Flagship and of the flotsam pick-up setting is setted to either 'off' or 'escorts only', it will pass
+	if(( collector == player.Flagship() && flotsamSetting == Preferences::Flotsam::OFF ) || ( collector == player.Flagship() && flotsamSetting == Preferences::Flotsam::ESCORT ))
+		return;
+	// If the collector is one of the player's ships and it is not the player flagship and the flotsam pick-up setting is setted to either 'off' or 'flagship only', it will pass
+	if(( collector->IsYours() && collector != player.Flagship() && flotsamSetting == Preferences::Flotsam::OFF ) || ( collector->IsYours() && collector != player.Flagship() && flotsamSetting == Preferences::Flotsam::FLAGSHIP ))
 		return;
```
[the remark this answers is not in the document]
Comment (OcelotWalrus) [reply source/Engine.cpp:2270]: Sounds great! I actually learned C++ some weeks ago so I still need to learn more.

**Comment by OcelotWalrus** [comment]
> Thanks for making it a single setting.
> 
> It's good to maintain the same order of settings in UI and code. Or at least don't put flotsam-related settings between autoaim and autofire.

I've done everything you requested.

**Comment by OcelotWalrus** [comment]
Yeah it seems to work correctly!

**Comment by TomGoodIdea** [comment]
Could you finish reordering in Preferences.cpp, Preferences.h and PreferencesPanel.cpp?

**Comment by OcelotWalrus** [comment]
Sure sorry for my mistake.
Check https://github.com/endless-sky/endless-sky/pull/9007/commits/9b37e88f57f8de5e93186003b426ddb776178333. I made so the flotsam was after 'Expand ammo' when it was possible and when it wasn't, after 'Boarding priority'. Is that good for oy?

**Comment by OcelotWalrus** [comment]
It's actually not really organized in the code.

File: source/Preferences.cpp [hunk 2]
```diff
@@ -479,8 +484,6 @@ const string &Preferences::AutoAimSetting()
 	return AUTO_AIM_SETTINGS[autoAimIndex];
 }
```
Comment (TomGoodIdea) [inline source/Preferences.cpp:486]:
```suggestion



```

File: source/Preferences.cpp [hunk 3]
```diff
@@ -525,8 +528,25 @@ const string &Preferences::BoardingSetting()
...
+	return static_cast<Flotsam>(flotsamIndex);
+}
+
 
 
+const string &Preferences::FlotsamSetting()
+{
+	return FLOTSAM_SETTINGS[flotsamIndex];
+}
+
```
Comment (TomGoodIdea) [inline source/Preferences.cpp:549]:
```suggestion



```

File: source/Preferences.cpp [hunk 4]
```diff
@@ -525,8 +528,25 @@ const string &Preferences::BoardingSetting()
 	return BOARDING_SETTINGS[boardingIndex];
 }
```
Comment (TomGoodIdea) [inline source/Preferences.cpp:530]:
```suggestion



```

File: source/PreferencesPanel.cpp [hunk 5]
```diff
@@ -62,6 +62,7 @@ namespace {
 	const string STATUS_OVERLAYS_ENEMY = "   Show enemy overlays";
 	const string STATUS_OVERLAYS_NEUTRAL = "   Show neutral overlays";
 	const string EXPEND_AMMO = "Escorts expend ammo";
+	const string FLOTSAM_SETTINGS = "Flotsam collection";
```
Comment (TomGoodIdea) [inline source/PreferencesPanel.cpp:65]:
```suggestion
	const string FLOTSAM_SETTING = "Flotsam collection";
```

File: source/PreferencesPanel.cpp [hunk 6]
```diff
@@ -243,6 +244,8 @@ bool PreferencesPanel::Click(int x, int y, int clicks)
 				Preferences::ToggleAutoFire();
 			else if(zone.Value() == EXPEND_AMMO)
 				Preferences::ToggleAmmoUsage();
+			else if(zone.Value() == FLOTSAM_SETTINGS)
```
Comment (TomGoodIdea) [inline source/PreferencesPanel.cpp:247]:
```suggestion
			else if(zone.Value() == FLOTSAM_SETTING)
```

File: source/PreferencesPanel.cpp [hunk 7]
```diff
@@ -561,8 +564,8 @@ void PreferencesPanel::DrawSettings()
 		TURRET_TRACKING,
 		TARGET_ASTEROIDS_BASED_ON,
 		BOARDING_PRIORITY,
-		"Flagship flotsam collection",
 		EXPEND_AMMO,
+		FLOTSAM_SETTINGS,
```
Comment (TomGoodIdea) [inline source/PreferencesPanel.cpp:568]:
```suggestion
		FLOTSAM_SETTING,
```

File: source/PreferencesPanel.cpp [hunk 8]
```diff
@@ -683,6 +686,11 @@ void PreferencesPanel::DrawSettings()
 		}
 		else if(setting == EXPEND_AMMO)
 			text = Preferences::AmmoUsage();
+		else if(setting == FLOTSAM_SETTINGS)
```
Comment (TomGoodIdea) [inline source/PreferencesPanel.cpp:689]:
```suggestion
		else if(setting == FLOTSAM_SETTING)
```

**Comment by TomGoodIdea** [comment]
I'd also suggest renaming the `Preferences::Flotsam` class to `Preferences::FlotsamSetting` or something, since we already have a `Flotsam` class and it might be confusing.

File: source/Preferences.cpp [hunk 9]
```diff
@@ -119,6 +119,9 @@ namespace {
 	const vector<string> BOARDING_SETTINGS = {"proximity", "value", "mixed"};
 	int boardingIndex = 0;
 
+	const vector<string> FLOTSAM_SETTINGS = {"off", "on", "flagship only", "escorts only"};
+	int flotsamIndex = 2;
```
Comment (TomGoodIdea) [inline source/Preferences.cpp:123]:
```suggestion
	int flotsamIndex = 1;
```
it should be "on" by default

**Comment by OcelotWalrus** [comment]
> Preferences::Flotsam

Doing it right now.

**Comment by OcelotWalrus** [comment]
Done, https://github.com/endless-sky/endless-sky/pull/9007/commits/343cc72ae50f4f4b317aed2b4511a0283d12c593. Also you changed FLOTSAM_SETTINGS TO FLOTSAM_SETTING at PreferencesPanel.cpp but not at Preferences.cpp, do I dit it.

**Comment by OcelotWalrus** [comment]
oops I made errors. Let me fix it

**Comment by TomGoodIdea** [comment]
FLOTSAM_SETTINGS is a vector with possible values and FLOTSAM_SETTING is the preference title.

**Comment by OcelotWalrus** [comment]
> FLOTSAM_SETTINGS is a vector with possible values and FLOTSAM_SETTING is the preference title.

So do I keep FLOTSAM_SETTINGS or FLOTSAM_SETTING at Preferences.cpp?

**Comment by OcelotWalrus** [comment]
Okay sorry.

**Comment by OcelotWalrus** [comment]
I think I need help, I renamed the class Flotsam to FlotsamSetting and I keep getting error and I can't resolve them. It says that in Preferences.cpp in the line 542, FlotsamSetting in class Preferences does not a name type.

**Comment by OcelotWalrus** [comment]
Wait I think I've figured it out:
image
It's because there is two FlotsamSetting (the class and the const).

**Comment by OcelotWalrus** [comment]
Nice no more error. Now I'm gonna test it again to be sure it still works.

**Comment by OcelotWalrus** [comment]
It still works.

**Comment by OcelotWalrus** [comment]
Also, right above the code that check for the player FlotsamCollectionSetting, I added just one comment:
image

File: source/Preferences.h [hunk 10]
```diff
@@ -132,6 +139,11 @@ class Preferences {
 	static BoardingPriority GetBoardingPriority();
 	static const std::string &BoardingSetting();
 
+	// Flotsam setting, either "off", "on", "flagship only", or "escorts only".
+	static void ToggleFlotsam();
+	static FlotsamCollection GetFlotsam();
```
Comment (TomGoodIdea) [inline source/Preferences.h:144]:
```suggestion
	static FlotsamCollection GetFlotsamCollection();
```

File: source/Preferences.cpp [hunk 11]
```diff
@@ -527,6 +532,27 @@ const string &Preferences::BoardingSetting()
...
 
 
+void Preferences::ToggleFlotsam()
+{
+	flotsamIndex = (flotsamIndex + 1) % FLOTSAM_SETTINGS.size();
+}
+
+
+
+Preferences::FlotsamCollection Preferences::GetFlotsam()
```
Comment (TomGoodIdea) [inline source/Preferences.cpp:542]:
```suggestion
Preferences::FlotsamCollection Preferences::GetFlotsamCollection()
```

File: source/Engine.cpp [hunk 12]
```diff
@@ -2261,8 +2261,21 @@ void Engine::DoCollection(Flotsam &flotsam)
 	}
 	if(!collector)
 		return;
-	if(collector == player.Flagship() && !Preferences::Has("Flagship flotsam collection"))
-		return;
+
+	// Checks for player FlotsamCollection setting
+	if(collector->IsYours())
+	{
+		const auto flotsamSetting = Preferences::GetFlotsam();
```
Comment (TomGoodIdea) [inline source/Engine.cpp:2268]:
```suggestion
		const auto flotsamSetting = Preferences::GetFlotsamCollection();
```

**Review by TomGoodIdea — APPROVED** [review]

**Comment by OcelotWalrus** [comment]
Thanks for approving.

**Review by Hurleveur — APPROVED** [review]
The code looks good, nice reviewing there @TomGoodIdea !
Conceptually expending that setting makes sense as well.

**Comment by OcelotWalrus** [comment]
Thanks!

**Pull request merged**